rfcs/2022-12-09-541-reload-api.md

```diff
@@ -0,0 +1,208 @@
+
+The control server will support a single route at `POST /config`, which will
+accept a JSON-encoded version of a Vector configuration. Similar to config
+received from a provider implementation, input globs will be expanded but
+environment variables will not be interpolated. The HTTP response will
+synchronously indicate the success or failure of the triggered reload,
+including any error messages scoped by component name.
+
+### Implementation
+
+The goal will be to implement a very simple, local-only API via which external
+processes can hook directly into Vector's reloading process. An important
+difference with existing solutions like providers is that the interaction will
+be synchronous, i.e. there will be a response with either success or failure
+and error messages for each reload attempt.
+
+Currently, all of Vectors reloading ultimately happens in the main run loop,
+where we watch for events like a signal, crash of a component, etc, and execute
+the corresponding changes to the topology (slightly different depending on the
+source of the signal). The first challenge will be to consolidate the logic of
+what needs to happen when a config changes (e.g. updating GraphQL API state,
+enterprise reporting, internal events, etc). A rough initial pass at that
+process is included with the RFC PR to help show dependencies. The next step
+would likely be to extract a `TopologyController` or similar struct that owns
+the topology and everything that needs to change along with it. That struct can
+then present a simple API for reloads that ensures everything is done in a
+consistent matter, no matter the source.
+
+Once that struct is in place, it can be wrapped in an `Arc<Mutex<_>>` or
+similar such that it can be shared across both the existing `Signal`-focused
+run loop and a new control server. The new control server itself can be a very
+simple task spawned prior to entering the main loop. As a local-only control
+API, it can different from most HTTP servers in a few ways:
+
+- By listening on a Unix socket instead of the network, it can rely on
+  filesystem permissions for authentication rather than its own scheme.
+- Because it's designed for use with a single local client, it does not need to
+  support concurrent processing of any kind. If it can't immediately acquire
+  the lock on the `TopologyController`, it can fail and tell the client to try
+  again.
+
+With the server and controller in place, the path for synchronous feedback will
+exist. The next step is to adjust the implementation of
+`RunningTopology::reload_config_and_respawn` away from the current return value
+of `Result<bool, ()>` and towards something that actually returns error
+messages. This will involve a fair bit of work down into the reloading stack,
+where we currently tend to emit errors as they happen. The likely solution will
+be refactoring to build up an error report of some kind that can be reported
+and/or returned at the end of the whole process.
```

Comment:
> would running in this mode need to implicitly disable, say, the signal-based control mechanism, or at least parts of it like disabling config watching?

I mentioned a more isolated mode as an alternative, but what's proposed here doesn't change any of this. Basically, I think it will take some dedicated effort to unravel those other threads. It's likely that we'll want to clean them up, but it doesn't need to block this work.

> Obviously asynchronous code can use a standard mutex without immediately deadlocking, but I do wonder if having the existing core loop need to go through that would make it possible for the local reloading API to somehow interact poorly with it.
> 
> Said another way: could this just be a command channel so that there's implicit serialization?

I thought about this initially but I don't think we actually want any "depth" to the channel, so it essentially boils down to the equivalent of a mutex. The existing stream of signals is already a stream, so it's possible they could stack up even in Vector's current state. But with the new server, I'd rather fail the request if a reload is already in progress than queue them up internally. My thought was that we'd use a tokio-flavored mutex around the controller to let the signal-handling loop wait when needed, and the server itself would just `try_acquire` or return that it's busy.

Reply:
As far as the channel vs mutex thing: yeah, that's reasonable.

But just to be clear in terms of the loop stuff: are you saying that as-is, ignoring the "more isolated reloading API" approach, we'd simply be letting this reloading API interact with the topology controller vs a lock, while the current signal handling loop also interacted with it via the topology controller?

If so, then I'd still want to re-ask the question: are we sure can ignore the interactions between the two in the first pass? Do we really not need to disable the portions of the current logic that try to reload existing config files, etc?

Like technically, sure, we could do that, but shouldn't the RFC really include the isolation aspect to make the config reloading API have more teeth? Like it does strike me that this is really meant to be more of an exclusive mode, rather than simply augmenting the existing approach.

Reply:
Yeah, all of that is fair and I don't see my question as being a blocker to this work, or this RFC as described.

I mostly wanted to drive towards the practical aspect of making sure doing it this way didn't have a gaping hole we weren't seeing because we hadn't yet _thought_ about if such a hole exists simply because the out-of-scope bits stopped us from bothering to think about it at all.

From your answer, though, you've clearly thought about it, so I'm confident with your assessment. 👍🏻